### DISCUSSION TITLE
AvaloniaUI source changes - Related to 3D implementation

### DISCUSSION THREAD

**Discussion by markusalbrecht-procam** [AvaloniaUI/Avalonia#14151] [General]
Hi Avalonia Team,

We are in the process of integrating OpenES via Angle into AvaloniaUI to facilitate the 3D component of our application. During this implementation, we encountered various issues that necessitated modifications to the AvaloniaUI source code. The specific challenges and corresponding adjustments are outlined below for your consideration:

**OpenGL Context Usage in Separate Applications:**
We observed an impediment in utilizing the OpenGL context in a distinct application. Consequently, we made alterations to the AvaloniaUI source code to address this issue. The root cause remains unclear; however, it is plausible that initialization and configuration within AvaloniaUI may be pivotal.

**Rendering 3D Models from STL Files:**
With the modified AvaloniaUI and integration of code segments from the "OpenGL Control" example, we successfully achieved the loading and display of 3D models from STL files (formatted in a clock version). Notably, the AvaloniaUI source code lacked bindings for certain functions, such as glDisable, and specific constants. The existing version can accommodate models with a polygon size of less than 65k triangles.

In light of these modifications, we are deliberating the best course of action to incorporate these changes into the official AvaloniaUI repository. Our primary concern is maintaining compatibility with future AvaloniaUI updates while ensuring the seamless integration of OpenES via Angle. Consequently, we are considering creating a Pull Request (PR) to contribute these enhancements.

We seek your guidance on the most appropriate approach to proceed with these changes. Our aim is to contribute to the AvaloniaUI ecosystem while preserving the ability to effortlessly upgrade AvaloniaUI NuGets in response to any subsequent developments on your end. Your insights on the optimal strategy for incorporating these modifications and any associated considerations would be highly valued.

Thank you for your time and consideration. We look forward to your guidance on this matter.

BR
Markus

**Comment by timunie** [comment]
@markusalbrecht-procam I cannot agree or disagree on any code changes here. But if you have done the needed changes already on your side, you can probably open a draft PR. That would make it possible for others to take a first look into the changes and some first feedback. Also it would generate packages to test the changes.

**Reply by markusalbrecht-procam** [reply]
@timunie Hi Tim, ok we will prepare a changeset and detailed description of the changes and then create a PR for those to start the further discussion and review.

**Comment by BAndysc** [comment]
(note: I am just a regular Avalonia user, not a dev)

Regarding 2. I think this is still valid and no more bindings should be added to the GlInterface:
> [kekekeks](https://github.com/kekekeks) [on May 21, 2021](https://github.com/AvaloniaUI/Avalonia/discussions/5959#discussioncomment-769129)
> GlInterface is meant for Avalonia internal purposes, not for user-facing OpenGL bindings. So there will be as less functions there as possible

There is an example of using Silk.NET with Avalonia https://github.com/kekekeks/Avalonia-Silk.NET-Example/tree/master/AvaloniaSilkExample, which contains every binding you need.

**Reply by markusalbrecht-procam** [reply]
@BAndysc Ok, yes SILK.Net sounds promising we will check that as another option - thanks for info.

**Reply by timunie** [reply]
In case that is a valid option, we may want to add a sample to Avalonia.Samples loading and showing STL files. Let me know your feedback here.

**Reply by markusalbrecht-procam** [reply]
Thank you for the offer. For now this is not needed as we already implemented the loading and some transformation already by doing some small adjustments to the source. The moment I know all the details what we changed and why I will get back to you, maybe we went into the wrong direction and that lead to the maybe unnecessary code changes ;)- I will get detailed info soon.

**Comment by maxkatz6** [comment]
> We observed an impediment in utilizing the OpenGL context in a distinct application. Consequently, we made alterations to the AvaloniaUI source code to address this issue.

Our OpenGLControl is supposed to be a simple and straightforward control for rendering OpenGL content. 
It's still possible to implement a custom open gl control with much more control, not limited by OpenGL alone as well.
We have samples for D3D and Vulkan - https://github.com/AvaloniaUI/Avalonia/tree/master/samples/GpuInterop

It basically all starts from Compositor.TryGetCompositionGpuInterop as you can see in these samples. If there is any limitation of these gpu interop APIs - it should be improved instead of cloning repository. 

OpenGlControlBase itself is just a specific control implementation that built on top of this API.

**Reply by markusalbrecht-procam** [reply]
Thanks for the info it turned out that I got some wrong information from the 3D programmer in fact most of the changes where just modifications in the sample itself where we will anyways use our own stuf later on.

In the OpenGL sources we changed the following (in fact no big changes):
Avalonia\src\Avalonia.OpenGL\GlInterface.cs
![image](https://github.com/AvaloniaUI/Avalonia/assets/56921919/885d50b0-2aae-43f3-968c-6cee1a159c5f)

Avalonia\src\Avalonia.OpenGL\GlConsts.cs
![image](https://github.com/AvaloniaUI/Avalonia/assets/56921919/395e4f5b-22ab-4516-be81-3793fe054f5c)
![image](https://github.com/AvaloniaUI/Avalonia/assets/56921919/52846275-b0ba-462b-bc5e-be10ecaac277)
![image](https://github.com/AvaloniaUI/Avalonia/assets/56921919/23808a0d-ac2b-4333-8f82-1f6c59f30edc)
![image](https://github.com/AvaloniaUI/Avalonia/assets/56921919/edae0bf7-b60b-4385-8936-98bbd828b2dc)
![image](https://github.com/AvaloniaUI/Avalonia/assets/56921919/70c2472c-36dd-4cd3-a358-c6d85eaf5b6a)

This was the reason for it:
glEnable(param)/glDisable(param) - enable/disable rasterizer, scissor test, stencil test, depth test, alpha blending states on GPU
constant from GlConsts.cs - it's a simple copy paste from OpenGL ES 3 header file GLES3/gl3.h 


Maybe that's the wrong approach and we should integrate different ?

Open for any input. br Markus

**Reply by maxkatz6** [reply]
As it was mentioned before, GlInterface is implemented specifically for the Avalonia needs. If you need custom members, it shouldn't be difficult to use them. Silk.NET is a typical solution. Avalonia GLInterface provides GetProcAddress that allows to get a pointer to any GL method.

**Comment by markusalbrecht-procam** [comment]
We finall ymaged to get that thing running at least a bit - now we are encountering other issues therefor I will create a new Discussion.
